File: problems/0208-implement-trie-prefix-tree/python/solution.py
class Trie:
    """Implement Trie (Prefix Tree) — https://leetcode.com/problems/implement-trie-prefix-tree/"""

    def __init__(self):
        self.root = {}

    def insert(self, word):
        node = self.root
        for c in word:
            node = node.setdefault(c, {})
        node["$"] = True

    def _find(self, s):
        node = self.root
        for c in s:
            if c not in node:
                return None
            node = node[c]
        return node

    def search(self, word):
        node = self._find(word)
        return node is not None and "$" in node

    def startsWith(self, prefix):
        return self._find(prefix) is not None


def trieOps(ops, args):
    """Test driver: replay LeetCode-style operation/argument lists."""
    trie = Trie()
    out = []
    for op, arg in zip(ops, args):
        if op == "Trie":
            trie = Trie()
            out.append(None)
        elif op == "insert":
            trie.insert(arg[0])
            out.append(None)
        elif op == "search":
            out.append(trie.search(arg[0]))
        elif op == "startsWith":
            out.append(trie.startsWith(arg[0]))
    return out

File: problems/0208-implement-trie-prefix-tree/typescript/solution.ts
// Implement Trie (Prefix Tree) — https://leetcode.com/problems/implement-trie-prefix-tree/
class TrieNode {
  children = new Map<string, TrieNode>();
  end = false;
}

class Trie {
  private root = new TrieNode();

  insert(word: string): void {
    let node = this.root;
    for (const c of word) {
      let next = node.children.get(c);
      if (next === undefined) {
        next = new TrieNode();
        node.children.set(c, next);
      }
      node = next;
    }
    node.end = true;
  }

  private find(s: string): TrieNode | null {
    let node: TrieNode = this.root;
    for (const c of s) {
      const next = node.children.get(c);
      if (next === undefined) return null;
      node = next;
    }
    return node;
  }

  search(word: string): boolean {
    const node = this.find(word);
    return node !== null && node.end;
  }

  startsWith(prefix: string): boolean {
    return this.find(prefix) !== null;
  }
}

export function trieOps(ops: string[], args: string[][]): (boolean | null)[] {
  let trie = new Trie();
  const out: (boolean | null)[] = [];
  for (let i = 0; i < ops.length; i++) {
    const arg = args[i];
    switch (ops[i]) {
      case "Trie":
        trie = new Trie();
        out.push(null);
        break;
      case "insert":
        trie.insert(arg[0]);
        out.push(null);
        break;
      case "search":
        out.push(trie.search(arg[0]));
        break;
      case "startsWith":
        out.push(trie.startsWith(arg[0]));
        break;
    }
  }
  return out;
}
